### ISSUE TITLE
Does test_minibatch optimize model parameters or just forward?

### ISSUE THREAD

**Issue by Yogurt928** [Microsoft/CNTK#2476]
Hi,
I want to train a binary classify and modified the FeedForward script from the tutorial to deal with my own data. The script can run normally but the result seems very strange, detail can be seen as following:
1. About data
I have 2000 files, there is 1600 samples in each file. Each sample is a  vector with 5510 dimensions (5508 for feature, 2 for label). Because of the data size is too large to load in memory one time, the file is saved as binary format and I will process it one file by one file. I use 1600 files as the training data and 400 files as the testing data.
```Python
def read_records(format, f):
    record_struct = Struct(format)
    chunks = iter(lambda: f.read(record_struct.size), b'')
    return (record_struct.unpack(chunk) for chunk in chunks)
def load_data_sample_bin(featFile):	
	print("load bin...")
	f=open(featFile,'rb')
	features = np.zeros((chunkSize,input_dim))
	labels = np.zeros((chunkSize,num_output_classes))
	i=0
	for rec in read_records('<5510f', f):
		rec=list(rec)
		feat=np.array([rec[:input_dim]])
		lab=np.array([rec[input_dim:len(rec)]])
		features[i]=feat
		labels[i]=lab
		i=i+1
	features=features.astype(np.float32)
	labels=labels.astype(np.float32)
	return features, labels
```

2. About model
I used a feedforward network with 2 hidden layers and  16 nodes for each layer.
```Python
num_hidden_layers = 2
hidden_layers_dim = 16
input_dim=2754*2
num_output_classes=2

def create_model(features):
	with default_options(init=glorot_uniform(), activation=C.sigmoid):
		h = features
		for _ in range(num_hidden_layers):
			h = Dense(hidden_layers_dim)(h)
		last_layer = Dense(num_output_classes, activation = None)

		return last_layer(h)

input = C.input_variable(input_dim)
label = C.input_variable(num_output_classes)

z = create_model(input)
loss = C.cross_entropy_with_softmax(z, label)
eval_error = C.classification_error(z, label)
lr_schedule = learning_rate_schedule(lr_per_sample, UnitType.sample, epoch_size)
learner = sgd(z.parameters, lr_schedule)
trainer = Trainer(z, (loss, eval_error), [learner])
```

3. About training
Each time, I will load one file from the training dataset, and use it as one minibatch to train the model. After training by all of the 1600 files, I will use the rest 400 files to do a testing. During the training, I will use one of the test file to get a test error after each minibatch training, and the training loss/traing error/test error will be print for each minibatch.
```Python
def print_training_progress(trainer, mb, frequency, features_test, labels_test, outputfile, verbose=1):    
	training_loss = "NA"
	eval_error = "NA"
	test_error = "NA"
	#print("print_training_progress")
	if mb%frequency == 0:
		#training_loss = get_train_loss(trainer)
		#eval_error = get_train_eval_criterion(trainer)
		training_loss = trainer.previous_minibatch_loss_average
		eval_error = trainer.previous_minibatch_evaluation_average
		test_error=trainer.test_minibatch({input : features_test, label : labels_test})
		thistime=time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())
		if verbose: 
			print ("[{}], Minibatch: {}, Train Loss: {}, Train Error: {}, Test Error: {}".format(thistime, mb, training_loss, eval_error, test_error), file=outputfile)
		
	return mb, training_loss, eval_error, test_error

chunkSize=1600
minibatch_size = 1600
epoch_size = 20 
lr_per_sample = [0.0001,0.0001]

# [Temp testing data]
features_test_temp="NA"
labels_test_temp="NA"
featFile = os.path.join(featPath, 'feat_'+str(i*testChunkNum+1))
features_test_temp, labels_test_temp = load_data_sample_bin(featFile)

# [Trainning data]
trainChunkNum=1600
testChunkNum=400
chunkSize=1600

batchNum=1
for e in range(0, epoch_size):
     for j in (0, trainChunkNum):
	  features_train = "NA"
	  labels_train = "NA"
	  print('Training data: feat_'+str(j));
	  featFile = os.path.join(featPath, 'feat_'+str(j))
	  features_train, labels_train = load_data_sample_bin(featFile)

          # Run the trainer and perform model training
	  training_progress_output_freq = chunkSize/minibatch_size
	  feat = features_train[m*minibatch_size:(m+1)*minibatch_size,:]
	  lab = labels_train[m*minibatch_size:(m+1)*minibatch_size,:]
					
	  # [Training]
	  trainer.train_minibatch({input : feat, label : lab})
	  batchsize, loss, error, test_error = print_training_progress(trainer, batchNum, 										training_progress_output_freq,features_test_temp, labels_test_temp, outputfile, verbose=1)

          #[Evaluation / Testing](each epoch)
	  features_test="NA"
	  labels_test="NA"
	  for j in range(trainChunkNum, trainChunkNum+testChunkNum):
		print('Testing data: feat_'+str(j));
		featFile = os.path.join(featPath, 'feat_'+str(j))
		features_test, labels_test = load_data_sample_bin(featFile)
		test_error=trainer.test_minibatch({input : features_test, label : labels_test})
		acc=1-test_error
		print("Accuracy: {}".format(acc))

```

4. About output
The test result is very bad, average accuracy for each epoch is around 0.5~0.6.
But what makes me confused is that the training error/testing error after each minibatch training decreased with time as following (The first epoch):
[2017-10-11 14:48:26], Minibatch: 1, Train Loss: 0.722729873657, Train Error: 0.5, Test Error: 0.5
[2017-10-11 14:48:27], Minibatch: 2, Train Loss: 0.694538116455, Train Error: 0.5, Test Error: 0.5
[2017-10-11 14:48:28], Minibatch: 3, Train Loss: 0.693884124756, Train Error: 0.5, Test Error: 0.5
[2017-10-11 14:48:29], Minibatch: 4, Train Loss: 0.69243522644, Train Error: 0.5, Test Error: 0.5
[2017-10-11 14:48:30], Minibatch: 5, Train Loss: 0.69509803772, Train Error: 0.5, Test Error: 0.5
[2017-10-11 14:48:31], Minibatch: 6, Train Loss: 0.693714065552, Train Error: 0.5, Test Error: 0.5
[2017-10-11 14:48:32], Minibatch: 7, Train Loss: 0.69256111145, Train Error: 0.5, Test Error: 0.5
[2017-10-11 14:48:33], Minibatch: 8, Train Loss: 0.693338241577, Train Error: 0.5, Test Error: 0.495625
......
[2017-10-11 15:15:18], Minibatch: 1597, Train Loss: 0.606590690613, Train Error: 0.304375, Test Error: 0.31375
[2017-10-11 15:15:19], Minibatch: 1598, Train Loss: 0.653539962769, Train Error: 0.280625, Test Error: 0.325625
[2017-10-11 15:15:20], Minibatch: 1599, Train Loss: 0.700058441162, Train Error: 0.360625, Test Error: 0.264375
[2017-10-11 15:15:22], Minibatch: 1600, Train Loss: 0.631870117187, Train Error: 0.285, Test Error: 0.269375

5. My confusion
If the model is bad, why the test error for each minibatch training can be reduced gradually?
In the first epoch, for each minibatch, the training file is new to the model, why it decreased with time too? Is there something wrong in my code?
I put the mainly code as above, if there is anything not clear, please tell me.
Hope for advice, thanks!

**Comment by cha-zhang** [comment]
Your model looks fine. Have you tuned the learning rate, or increase the hidden dimension?

**Comment by Yogurt928** [comment]
@cha-zhang 

Thanks for reply!
I have found the reason why the performance is so bad. 
Because of my careless, I take a log operation on the training data but forget to do that on the testing data... (/ω＼)

Thanks again!
